### PR TITLE
Add anki extension

### PR DESCRIPTION
**Pull request by anton-suprun**
## Description

Anki client for Raycast. Offers a subset of Anki features through the powerful and rich Raycast APIs.

Supporting the following features:

| Command Title | Description | Supported Actions |
|-------|-------------|-------------------|
| Decks | Lists all Anki decks | Create Deck, Browse Deck, Study Deck, Delete Deck, Add Card To Deck, Sync with Web |
| Browse Cards | Browse all cards in Anki | Search, View, View Metadata, Delete, View Files in Card |
| Add Card | Add a card to a deck in Anki | Create, Clear Form |
| View Stats | View full collection stats | - |

*Note: Several of the actions could be re-purposed as a command however my reasoning for utilizing actions instead of commands is to avoid cluttering a user's Raycast with too many commands (esepcially where an Action is sufficient ) and because I want to try my best to have the UI mimic/emulate the Anki UI so that users are more familiar with it.*

Example of what I mean by that: syncing the collection with the web version is done from the decks view in Anki UI; I have added an Action `sync` in the Decks command.

## Screencast

*Note: there are several screencasts here, each displaying various features of the applications (actions, toasts, etc). The file title more-or-less describes the features I'm showing in the video*

Please let me know if additional screencasts are required or if you spot any issues/concerns 🙏 

https://github.com/user-attachments/assets/4c292127-4cd2-42a6-ad42-9df894394684

https://github.com/user-attachments/assets/0106857e-e64e-4adb-983f-c413a55c002b

https://github.com/user-attachments/assets/3e21705f-12f6-46d0-b9e4-1b33a35910dd

https://github.com/user-attachments/assets/e0a004a0-29a2-4eac-9cd1-6740527c0b85

https://github.com/user-attachments/assets/09ea6b04-da77-43c7-8b23-2e06efb3d5dd

https://github.com/user-attachments/assets/c15ef9e8-cabb-4e50-a79c-f98a6509b013

## Checklist

- [x] I read the [extension guidelines](https://developers.raycast.com/basics/prepare-an-extension-for-store)
- [x] I read the [documentation about publishing](https://developers.raycast.com/basics/publish-an-extension)
- [x] I ran `npm run build` and [tested this distribution build in Raycast](https://developers.raycast.com/basics/prepare-an-extension-for-store#metadata-and-configuration)
- [x] I checked that files in the `assets` folder are used by the extension itself
- [x] I checked that assets used by the `README` are placed outside of the `metadata` folder

### CODE REVIEW

**Comment by koinzhang** [comment]
There seems to have been a similar PR before. https://github.com/raycast/extensions/pull/13130

**Comment by anton-suprun** [comment]
> There seems to have been a similar PR before. #13130

Ah it would appear so; another Anki user 😊 

Myabe we can collaborate on this; there still so many features I plan to add so having help would be awsome. I would be happy to reach out to this user and ask if they'd consider collaborating. 

After looking their source code and their supported features; it would appear though that currently my extension supports significatly more critical features that I don't see in their extension:

- Action to study the cards; this is the *key* feature of this extension. It utilizes Anki's ranking algorithm under the hood.
- Show useful statistics such as: deck stats (due, learn, new, total), card metadata (
- Support more basic note types (there are many other); I currently support all but Occlusion,Cloze and Input types (which I'm currently still working on but they require some more complex logic)
- Have a command for browsing cards or searching cards; I have support for searching utilizing the familiar Anki syntax. This means support for searching by tags, decks, content and much more.
- Support for files (neither uploading nor viewing); I have a features to view files (in Grid), quick previev, seach by filename, sort by field. Users will often upload images to help them in their studies for example.
- option to trigger GUI Anki actions. With the Anki, users may occasionally want to jump in to the application itself. These actions allow the user to jump into the GUI app; to the action they were doing in the extension. So if a user is browsing cards; triggering the GUI Action, will open the browser dialog in Anki 
    
**_How do you normally proceed with cases as such?_** 

Looking through the other CR I noticed  there has been no activity for almost a month.

Also, it would appear my initial commit are about a [month earlier](https://github.com/anton-suprun/anki-raycast/commits/main/), (June 1st) in contrast to theirs which is on [June 28)(https://github.com/AlpayC/super-anki-raycast-extension/commits/main/).

**Comment by anton-suprun** [comment]
> There seems to have been a similar PR before. #13130

I've included some screencasts to demo some the features I [mentioned](https://github.com/raycast/extensions/pull/13616#issuecomment-2245167492)

**Comment by pernielsentikaer** [comment]
Are you still around, @AlpayC? Would you be up for a collaboration with @anton-suprun

**Comment by AlpayC** [comment]
@pernielsentikaer how is the pr order? Our extension was published first, so i wonder, if i will stay as the author, when I bring up my old pr for anki support

**Comment by pernielsentikaer** [comment]
There is no specific order since there was no activity in #13130, so it got closed by stalebot. The best approach would be to take the best aspects of both extensions and unite them into one

**Comment by AlpayC** [comment]
> There is no specific order since there was no activity in #13130, so it got closed by stalebot. The best approach would be to take the best aspects of both extensions and unite them into one

So the question remains as to who will remain the author of the extension if they work together.

**Comment by pernielsentikaer** [comment]
There can be one author but multiple contributors (and even past contributors)

**Comment by AlpayC** [comment]
> There can be one author but multiple contributors (and even past contributors)

Ok than i dont want to collberate

**Comment by AlpayC** [comment]
I also think it's a shame that raycast anticipates the possibility, even though they want to expand the possibilities for other users and had the idea first. We all know that PR sometimes takes longer. 

Now there's another developer who had the same idea but submitted the whole project too late. 
@pernielsentikaer

**Comment by pernielsentikaer** [comment]
I'm sorry that you feel that way. It's impossible for others to see that you had a PR with the same thing (which got closed due to inactivity). We wrote multiple times in the PR but received no response.

Maybe it's fine with @anton-suprun that you continue yours and he'll be a contributor instead?

**Comment by anton-suprun** [comment]
Hi, 

I would be happy to collaborate. Can we go ahead and use my extension though since it currently my extension supports significantly more features. Also, the pattern I'm using for calling anki-connect APIs is easily extensible so we can move faster. Would you consider contributing to my extension ?

On that note, yes, despite my project being published to Raycast later, the first commit to it was a month before yours

**Comment by AlpayC** [comment]
Sry @anton-suprun i dont want to collaborate since  i made the PR first. And i want to remain as the author. The same features of yours are in development and i have the anki server support as well in my version of the extension.

@pernielsentikaer i want to reopen my pr and fix the remaining points of my pr, so i can finalize it. Can you open it please?

**Comment by anton-suprun** [comment]
> Sry @anton-suprun i dont want to collaborate since  i made the PR first. And i want to remain as the author. The same features of yours are in development and i have the anki server support as well in my version of the extension.
> 
> @pernielsentikaer i want to reopen my pr and fix the remaining points of my pr, so i can finalize it. Can you open it please? 

I'm not sure I understand why you feel strongly about being the author; you'd be adding code for features  that already exists in this current PR; that seems counter productive to me.

Also, about the PR order; I can make the same argument about the commit order. I initially committed to my open repo a month before you. That combined with the fact that you were inactive for about a month when in review. I understand that sometimes it can take time but that is significant time for something you seemingly feel strongly about. 

I'm a huge fan of what Anki community has done and just want to contribute to it. I use my extension every day and I'm going to pro actively maintain it. Can you commit to that as well? 

I just wanted an extension in the Raycast store for others to use and as it currently stands, my extension has more features and an active maintainer. 

I'm not sure what decision Raycast will decide. I made this extension because I really saw potential in Raycast UI with Anki; and I've been an avid Anki user for a long time.

**Comment by AlpayC** [comment]
Well the same arguments can I list here too, but i think we need a decision from @pernielsentikaer and his team.

**Comment by anton-suprun** [comment]
> Well the same arguments can I list here too, but i think we need a decision from @pernielsentikaer and his team.

Agreed; let's see what they propose. 

(Funny predicament we find ourselves in 😅. Rather unique situation I would imagine haha)

(Btw; nice to meet you @AlpayC 👋. If the timelines of the projects were different we'd probably be collaborators and not having this discussion lol)

**Comment by anton-suprun** [comment]
@pernielsentikaer 

Let us know what you decide :) 

As it stands I think it would be fair to proceed with my extension. 

- I had an earlier commit
- The other PR was closed due to inactivity 
- Currently, my extension has significantly more Anki features supported 

I'm happy to collaborate but also don't want to repeat the work I've already done on my extension (adding the same features again, just to a different extension with different patterns)

Let us know your decision

File: extensions/anki/assets/command-icon.png
Comment (LitoMore) [inline extensions/anki/assets/command-icon.png:1]: This file needs to be removed.

File: extensions/anki/notes.md [hunk 1]
```diff
@@ -0,0 +1,4 @@
+### Features
+
+- Configure Anki
+  - Need an option to cofigure the path to the Anki collection
```
Comment (LitoMore) [inline extensions/anki/notes.md:4]: Is this file still needed?

**Review by LitoMore — CHANGES_REQUESTED** [review]

File: extensions/anki/notes.md (hunk 1 above)
Comment (anton-suprun) [reply extensions/anki/notes.md:4]: Thanks for pointing this out. It is not; will remove now.

File: extensions/anki/assets/command-icon.png
Comment (anton-suprun) [reply extensions/anki/assets/command-icon.png:1]: Hi👋 

Thanks for having a look at this PR 🙂 

Sorry, I'm not sure I understand why this needs to be removed, this file is used as the icon for the extension. Could you please elaborate why it needs to be removed?

File: extensions/anki/notes.md (hunk 1 above)
Comment (anton-suprun) [reply extensions/anki/notes.md]: Removed ✅ 

Thanks!

File: extensions/anki/assets/command-icon.png
Comment (anton-suprun) [reply extensions/anki/assets/command-icon.png:1]: Removed ✅ 

Thanks!

**Comment by pernielsentikaer** [comment]
It's indeed unfortunate that multiple users are creating the same extension. Most of the time, it's possible to merge the best features from both into one extension, which provides the end users with the best possible experience.

I'm now done checking all the related PRs and comments, and this PR is the one we will continue with, as it was opened after the previous one was closed due to inactivity.

We always ping authors at least three times (at review, when the stalebot gives a warning about inactivity, and when it's getting closed).

The reason we need to have a stalebot is the large number of PRs we have to review in addition to managing the community, so it helps us keep our backlog clean and prioritize extensions that are under active development.

> (Does Raycast have any community guidelines for scenarios like this 😅 or is this the first time a situation like this has occurred haha?)

In the past, there have been a few occasions where up to three users created the same extension due to hype in our community

**Comment by anton-suprun** [comment]
> It's indeed unfortunate that multiple users are creating the same extension. Most of the time, it's possible to merge the best features from both into one extension, which provides the end users with the best possible experience.
> 
> 
> 
> I'm now done checking all the related PRs and comments, and this PR is the one we will continue with, as it was opened after the previous one was closed due to inactivity.
> 
> 
> 
> We always ping authors at least three times (at review, when the stalebot gives a warning about inactivity, and when it's getting closed).
> 
> 
> 
> The reason we need to have a stalebot is the large number of PRs we have to review in addition to managing the community, so it helps us keep our backlog clean and prioritize extensions that are under active development.
> 
> 
> 
> > (Does Raycast have any community guidelines for scenarios like this 😅 or is this the first time a situation like this has occurred haha?)
> 
> 
> 
> In the past, there have been a few occasions where up to three users created the same extension due to hype in our community

Thanks for the consideration. Looking forward to contributing to  the Raycast community!

**Comment by pernielsentikaer** [comment]
I don't have Anki but it would be cool with a more seamless way to show that you don't have it (or turned it on(?))

Could be cool with a custom List.EmptyView, extensions like [eagle](https://www.raycast.com/Yukai/eagle) test for the application if you need some inspiration

Raycast01082024-yJaW5Hn1 png@2x
Raycast01082024-Knad3IO1 png@2x

**Comment by anton-suprun** [comment]
> I don't have Anki but it would be cool with a more seamless way to show that you don't have it (or turned it on(?))
> 
> 
> 
> Could be cool with a custom List.EmptyView, extensions like [eagle](https://www.raycast.com/Yukai/eagle) test for the application if you need some inspiration
> 
> 
> 
> Raycast01082024-yJaW5Hn1 png@2x
> 
> Raycast01082024-Knad3IO1 png@2x
> 
> 

Ah yes that's a good point; when I added that toast, I was wondering what would be the best way to indicate to a user that Anki is not currently on. 

I will look into the example you provided and add a change 🙂 

Thank you!

**Comment by pernielsentikaer** [comment]
I think the example I gave you uses a toast, but it could very well look better with a List.EmptyView

**Comment by anton-suprun** [comment]
> I think the example I gave you uses a toast, but it could very well look better with a List.EmptyView

Acknowledged! Will make the change this evening (EST) :)

**Comment by anton-suprun** [comment]
> I think the example I gave you uses a toast, but it could very well look better with a List.EmptyView

Hi @pernielsentikaer,

I followed your suggestion and looked at how eagle does it. I like it; I added the following in the latest commit:

- Toast actions to install the necessary items (Anki and anki-connect)
- Error handling function (handles connection errors, anki errors and unknown errors -> which users are prompted to create a bug report)

Also, due to the nature of this integration (Anki via anki-connect); users may run into various issues and I want them to be able to self-resolve. Therefore, to indicate to the user that something is misconfigured, in case of a connection error to Anki, I've added specific handling. Please see example:

https://github.com/user-attachments/assets/e513d60f-1582-450f-8bc3-63d8419953a0


Also, when it is an Anki error; we display the error string to the user in the toast. Example:


https://github.com/user-attachments/assets/e2cd480c-765a-4885-9465-40270580b759

Let me know your thoughts on this approach to error handling :)

**Comment by pernielsentikaer** [comment]
It looks really good, should we consider moving to [useForm](https://developers.raycast.com/utilities/react-hooks/useform) for better handling of states, errors and validation?

**Comment by anton-suprun** [comment]
> It looks really good, should we consider moving to [useForm](https://developers.raycast.com/utilities/react-hooks/useform) for better handling of states, errors and validation?

Thank you 🙏 

Certainly, I'll update the AddCard component to utilize `useForm`, add validation and better error handling 👍

**Comment by anton-suprun** [comment]
> It looks really good, should we consider moving to [useForm](https://developers.raycast.com/utilities/react-hooks/useform) for better handling of states, errors and validation?

Hi @pernielsentikaer,

As per your suggestion:

- I've refactored the Add Card component to utilize `useForm`
- Added validation for input and file types
- Better error handling

I also included an additional change for setting configurations (if folks want to be able to create duplicate cards, they can configure that in the options nows)

#### Demo of the changes:

https://github.com/user-attachments/assets/b5cdf0a9-e99e-45f2-94f4-783b7ae5b528


https://github.com/user-attachments/assets/caa83dd4-aea1-4c91-8692-e07461bd8d39

Please let me know your thoughts 🙏 

Thanks!

**Comment by pernielsentikaer** [comment]
I think we're ready to add the first version to the store. Let's see how it goes 🙂

**Review by pernielsentikaer — CHANGES_REQUESTED** [review]
It looks like you made the screenshot yourself, could you make the [screenshots](https://developers.raycast.com/basics/prepare-an-extension-for-store#screenshots) with this tool to get margins right 🙂

**Comment by anton-suprun** [comment]
> It looks like you made the screenshot yourself, could you make the [screenshots](https://developers.raycast.com/basics/prepare-an-extension-for-store#screenshots) with this tool to get margins right 🙂

Hi @pernielsentikaer,

I've used the screenshot tool as advised. 

Also, there was an update (Anki and anki-connect). I updated and tested the extensions functionality. 
No breaking changes found 👍 

I'm happy to proceed with the initial release if you are.

**Review by pernielsentikaer — APPROVED** [review]
I think we're ready to merge the first version 🔥

**Comment by anton-suprun** [comment]
> I think we're ready to merge the first version 🔥 

Awesome! 🚀
Thank you 🙏
